### PR TITLE
LPS-103362 Migrates personal menu to react and Clay v3

### PR DESCRIPTION
**Pull request by jbalsas**

### CODE REVIEW

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+function PersonalMenu({itemsURL}) {
+	const [active, setActive] = useState(false);
+	const [items, setItems] = useState([]);
+
+	function preloadItems() {
+		if (!preloadPromise.current) {
+			preloadPromise.current = fetch(itemsURL)
+				.then(response => response.json())
+				.then(items => setItems(items));
```
Comment (wincent) [inline modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js]: It's not good practice to call a `set` method asynchronously if there's a chance that the component got unmounted before the callback fires (and even if you think it never will get unmounted, it is a brittle assumption). So you can guard this with a `isMounted` check:

```javascript
import {useIsMounted} from 'frontend-js-react-web`;
```

Then:

```javascript
const isMounted = useIsMounted();

// and in the then ....
then(items => {
  if (isMounted()) {
    setItems(items);
  }
});
```

In layout-content-page-editor-web I made [a `useStateSafe` hook](https://github.com/brianchandotcom/liferay-portal/blob/cd6ad8bc54f4a437abfc6fedbcfb338a877e6f0d/modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/core/hooks/useStateSafe.js) that can be used as a drop-in replacement for `useState` that bakes this check in. I think we just need to make sure the naming is good (I don't know that it is) and we can move it into frontend-js-react-web.

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		}
+
+		return preloadPromise.current;
+	}
+
+	function toggleActive(newVal) {
+		preloadItems().then(setActive(newVal));
```
Comment (wincent) [inline modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js]: Same issue with asynchrony here; need an `isMounted()` check.

Also, seems like good practice to always have a `catch()` somewhere; either here or in `preloadItems()` itself. I guess that in this case putting it in `preloadItems()` might be best; that way the user can try again by clicking:

```javascript
catch(error => {
  if (process.env.NODE_ENV === 'development') {
    console.error(error);
  }

  // Allow retry.
  preloadPromise.current = null;
});
```

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+			active={active}
+			alignmentPosition={Align.BottomRight}
+			items={items}
+			onActiveChange={toggleActive}
+			trigger={
+				<ClayButton
+					displayType="unstyled"
+					onFocus={preloadItems}
+					onMouseOver={preloadItems}
+					symbol="user"
```
Comment (wincent) [inline modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js]: Do we need this `symbol` here if we're drawing a `<ClaySticker />` + `<ClayIcon />` below with `symbol="user"` anyway? I am not familiar enough with the components to know. 😂

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js (hunk 1 above)
Comment (jbalsas) [reply modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js]: Thanks! I've heard this so many times and yet forgot to apply it! 😢

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js (hunk 2 above)
Comment (jbalsas) [reply modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js]: hhaha... yeah, I was just hacking this quickly and just cut some corners... will add this when @drewbrokke sends it back

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js (hunk 3 above)
Comment (jbalsas) [reply modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js]: Certainly not. I initially had `<ClayButtonWithIcon symbol="user">`... which isn't exactly what we needed. I forgot to remove it when I added the sticker+icon combo

**Comment by drewbrokke** [comment]
Just started reviewing :)

**Comment by drewbrokke** [comment]
Screen Shot 2019-10-18 at 10 00 59 AM

(Old on the left, React impl on the right)
Looks like a couple of style differences are present:
- The overflow changed, so now there's a scrollbar
- The icon itself is shifted up slightly so that it is not centered anymore

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/page.jsp [hunk 4]
```diff
@@ -45,98 +45,21 @@ String label = (String)request.getAttribute("liferay-product-navigation:personal
...
-													menu.style.top = bodyRegion.top + 'px';
-												}
-											}
-										);
-
-										this.refs.dropdown.refs.portal.on(
-											'rendered',
-											function(event) {
-												if (dropdown.expanded) {
-													this.element.classList.add('dropdown-menu-personal-menu');
```
Comment (drewbrokke) [inline modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/page.jsp]: I think we'll need to retain this css class on the component for the overflow styling to work correctly.

Comment (jbalsas) [reply modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/page.jsp]: I'll take a look at the styling differences with @matuzalemsteles and @pat270.

We're also missing the compass icon, which I think just needs a change in the model

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		return preloadPromise.current;
+	}
+
+	function toggleActive(newVal) {
+		preloadItems().then(setActive(newVal));
+	}
+
+	return (
+		<ClayDropDownWithItems
+			active={active}
```
Comment (drewbrokke) [inline modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js]: What does this do when passing it to `ClayDropDownWithItems`? It looks like a boolean, but unless I'm mistaken the `active` attribute is relevant to the items in the list.

File: modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+			alignmentPosition={Align.BottomRight}
+			items={items}
+			onActiveChange={toggleActive}
```
Comment (drewbrokke) [inline modules/apps/product-navigation/product-navigation-taglib/src/main/resources/META-INF/resources/personal_menu/js/PersonalMenu.es.js]: If `active` above can be removed, we can remove this too and probably the whole `active` state hook.

**Comment by jbalsas** [comment]
Closing in favour of https://github.com/drewbrokke/liferay-portal/pull/897

**Pull request closed without merging**